examples/test/qlib/CsvUtil/csvutil.qtest

```diff
@@ -39,14 +52,64 @@ UK,1234567893,\"Sony, Xperia S\",31052012";
 
         # Iterate through source
         CsvDataIterator i(CsvInput, opts);
-        list l = map $1, i;
-        testAssertion("CsvDataIterator 1", \equalsIterated(), (i, new ListIterator(l)));
+        #list l = map $1, i;
+        testAssertion("CsvDataIterator 1", \equalsIterated(), (new ListIterator(CsvRecords), i));
+
+        CsvStringWriter w( ("write-headers": False, "optimal-quotes" : True, "quote_escape" : '"')+opts );
+        string outstr = w.write(CsvRecords);
+        testAssertion("CsvDataWriter writer", \equals(), (CsvInput+"\n", outstr));  # ending CR is assymetrical when empty lines are not ignored
 
         # test with empty data and header lines
         i = new CsvDataIterator("", ("header-lines": 1));
         testAssertion("CsvDataIterator 2", \i.next(), (), RESULT_FAILURE);
     }
 
+    csvMultiTest() {
+        *hash opts = (
+            "fields": (
+                "type": "int",
+
+                "invoice_no": "string",
+                "date": ("type": "date", "format": "YYYYMMDD"),
+
+                "item_no": "string",
+                "item": "string",
+                "pcs": "int",
+                "price": "number",
+
+                "total": "number",
+            ),
+            "headers": (
+                "header": ("type", "invoice_no", "date"),
+                "item":   ("type", "item_no", "item", "pcs", "price"),
+                "footer": ("type", "total"),
+            ),
+            "resolve_type": softstring sub(hash values) {
+                softint type = exists values{'type'}?values{'type'}:values{'0'};
```

Comment:
That's too old documentation. Here's the latest: https://qore.org/manual/develop/lang/html/operators.html